mlapi/discriminating_algo.py

```diff
@@ -0,0 +1,129 @@
+import statistics
+
+class DiscriminatingAlgo(object):
+    # Discrimination parameters
+    facets_sample_count = 3  # minimum docs by facet to be accepted
+    tolerance_interval = 10 # max standard deviation accepted ()
+    nbr_facet_values = 2 # to consider a facet as having a large number of values
+
+    facets = [] # names
+    step2 = {}
+    step3 = {}
+
+    def discriminating_facets_algo(self, facets_by_document):
+        documents_by_facet = self.inverse_dictionary(facets_by_document)
+        data = self.get_formated_data(documents_by_facet)
+        self.execute_algorithm(data)
+        if not (len(self.step3)):
+            self.adjust_parameters()
+            self.execute_algorithm(data)
+        return self.get_values_by_facet(self.step3)
+
+    def inverse_dictionary(self, facets_by_document):
+        documents_by_facet = {}
+        for k, v in facets_by_document.items():
+            for value in v:
+                documents_by_facet[value] = documents_by_facet.get(value, [])
+                documents_by_facet[value].append(k)
+        return documents_by_facet
+
+    def get_formated_data(self, documents_by_facet):
+        data = {}
+        list = []
+        for facet, docs in documents_by_facet.items():
+            t = (facet.name, facet.value)
+            if t not in list:
+                tuple = (facet.name, facet.value)
+                data[tuple] = docs
+                list.append(tuple)
+            elif t in list:
+                data[tuple].append(docs)
+        return data
+
+    def get_facet_names(self, data):
+        list = []
+        for facet, docs in data.items():
+            if facet[0] not in list:
+                list.append(facet[0])
+        return list
+
+    def get_values_by_facet(self,data):
+        dict = {}
+        list = []
+        if data:
+            for facet, docs in data.items():
+                if facet[0] not in list:
+                    dict[facet[0]] = [facet[1]]
+                    list.append(facet[0])
+                elif facet[0] in list:
+                    dict[facet[0]].append(facet[1])
+        return dict
+
+    def get_facet_sample(self, data):
+        facet_values_by_facet_name = self.get_values_by_facet(data)
+        self.facets = self.get_facet_names(data)
+        cpt = 0
+        dictionary = {}
+        #count how many documents are in each facet
+        for index in range(len(self.facets)):
```

Comment:
you appear to be iterating over facets to get the index and facet element

Pythonic way to do this is:

for i, facet in enumerate(self.facets)

https://stackoverflow.com/questions/22171558/what-does-enumerate-mean

Reply:
Same for:

for index in range(len(facet_names)):
for i in range(len(fv)):